File: cpp/algorithms/graph/topological_sort/kahn.cpp
#include <bits/stdc++.h>
using namespace std;

optional<vector<int>> topological_sort(const vector<vector<int>>& adj) {
    int n = (int)adj.size();
    vector<int> indeg(n, 0);
    for (int u = 0; u < n; ++u)
        for (int v : adj[u]) ++indeg[v];
    queue<int> q;
    for (int i = 0; i < n; ++i)
        if (!indeg[i]) q.push(i);
    vector<int> order;
    while (!q.empty()) {
        int u = q.front();
        q.pop();
        order.push_back(u);
        for (int v : adj[u])
            if (--indeg[v] == 0) q.push(v);
    }
    if ((int)order.size() != n) return nullopt;
    return order;
}

int main() {
    vector<vector<int>> g(3);
    g[0] = {1, 2};
    g[1] = {2};
    auto t = topological_sort(g);
    assert(t && *t == vector<int>({0, 1, 2}));
    cout << "topological_sort OK" << endl;
    return 0;
}

File: python/algorithms/graph/topological_sort/kahn.py
"""Kahn 拓扑排序。"""

from __future__ import annotations
from collections import deque


def topological_sort(adj: list[list[int]]) -> list[int] | None:
    n = len(adj)
    indeg = [0] * n
    for u in range(n):
        for v in adj[u]:
            indeg[v] += 1
    q: deque[int] = deque([i for i in range(n) if indeg[i] == 0])
    order: list[int] = []
    while q:
        u = q.popleft()
        order.append(u)
        for v in adj[u]:
            indeg[v] -= 1
            if indeg[v] == 0:
                q.append(v)
    if len(order) != n:
        return None
    return order


if __name__ == "__main__":
    # 0->1->2, 0->2
    g = [[] for _ in range(3)]
    g[0].extend([1, 2])
    g[1].append(2)
    t = topological_sort(g)
    assert t is not None and t == [0, 1, 2]
    g_cycle = [[1], [0], []]
    assert topological_sort(g_cycle) is None
    assert topological_sort([[]]) == [0]
    print("topological_sort OK")
